### PR TITLE
Use threads for button callback

### PR DESCRIPTION
**Pull request by QuzarDC**
The current implementation of the button callback in controller.c merely calls the callback function. While this is very useful for simple things, the fact that it runs the callback during an interrupt causes a number of issues. It limits the scope of what such a callback can do, which has been quite exemplified by our standard usage of it to call arch_exit. 

The idea here is to raise the possibility of configuring the maple button callback routine to spin off a new thread with the function rather than run it entirely in the context of the interrupt that triggered the callback.

It may be the case that this should be an optional behavior alongside the current one. I'm not entirely sure the scope of usage of this functionality and how it could be negatively impacted by the change. I can confirm though that it works with a few different tests in multi-threaded environments that otherwise trigger various errors regarding mutex_lock or thd_join being called in an interrupt.

### CODE REVIEW

**Comment by gyrovorbis** [comment]
Cool! I wasn't even aware of the issues this was creating, but I can definitely see how being able to call this in more contexts could be useful. 

Can't really comment much on the side effects of the implementation or do much other than minor C suggestions ,but the solution looks purdy. :)

**Comment by QuzarDC** [comment]
@pcercuei (zcrc in discord) raised the notion that this may merely be a stopgap to the implementation of threaded interrupts. With that in mind, I reworked this to use a standing thread created when the callback in installed rather than being created for the interrupt. In addition, it now cleans up after itself and can be un-set by setting to NULL.

It raised the question though of 'why have this, and not other interrupt callbacks spin off a thread?'. Compared to the majority of interrupt handlers though, this functionality is presented as something to be integrated into what would normally be user-land code. It's possible that the SIP and snd_stream callbacks would be candidates for such a change, as they are similar in that respect. The DMA callbacks might, too, but they are all one step closer into the interrupt handler than the ones mentioned so far.

Finally, the behavior of it now schedules the thread next to run immediately. While this makes it very simple, it also doesn't give the scheduling benefits that just sending it to the queue might.

**Comment by QuzarDC** [comment]
As an update/clarification to the last comment, I wanted to add the three big questions I have about this change:
1)Do we want it? Perhaps we should just be noting that the callback happens in an interrupt, and shouldn't be used for stuff like arch_exit.
2)If we do, is this the right scope to set something like this up in? A thread could be used at the interrupt level or in a few places up the way.
3)If so, is this the right way to do it? This could be the template of redoing dma callbacks or the snd_stream one so getting the structure/layout/details right would be useful.

**Comment by QuzarDC** [comment]
Marking this back as 'ready for review'. The change should be entirely transparent to users but make the button callback work significantly closer to how it's expected to work. The only reason I had marked it as draft was to sort of push to get some more discussion before merging it as this is sort of a 'minimal simplest' version of the implementation, and we don't really have anything that works like it currently in the codebase.

File: kernel/arch/dreamcast/hardware/maple/controller.c
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/controller.c:35]:
```suggestion
static kthread_t *btn_callback_thd = NULL;
```

File: kernel/arch/dreamcast/hardware/maple/controller.c
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/controller.c:37]:
```suggestion
static uint8_t btn_callback_arg_addr = 0;
static uint32_t btn_callback_arg_btns = 0;
```

File: kernel/arch/dreamcast/hardware/maple/controller.c
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/controller.c:52]:
```suggestion
static void *btn_callback_wrapper(void *args) {
    (void)args;

```

File: kernel/arch/dreamcast/hardware/maple/controller.c
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/controller.c:65]:
```suggestion
    /* This means either the callback is shutting down the 
       whole system, or some jerk called this in the callback. */
    if(thd_get_current()->tid == btn_callback_thd->tid)
        return;
```

File: kernel/arch/dreamcast/hardware/maple/controller.c
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/controller.c:75]:
```suggestion
```

File: kernel/arch/dreamcast/hardware/maple/controller.c
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/controller.c:80]:
```suggestion
```

File: kernel/arch/dreamcast/hardware/maple/controller.c
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/controller.c:98]:
```suggestion
    thd_set_label(btn_callback_thd, "cont_btn_callback");
```

File: kernel/arch/dreamcast/hardware/maple/controller.c
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/controller.c:196]:
```suggestion
```

File: kernel/arch/dreamcast/hardware/maple/controller.c
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/controller.c:96]: It's actually not safe to call this with interrupts enabled, as it could corrupt the runqueue if a thread scheduling interrupt comes in in the middle of this... Let me ponder the best way to deal with this (probably with another attribute in the `kthread_attr_t` struct).

File: kernel/arch/dreamcast/include/dc/maple/controller.h
Comment (ljsebald) [inline kernel/arch/dreamcast/include/dc/maple/controller.h:227]: The part above this in the comment probably should get updated, since it says that the callback will be called in an interrupt context, which is no longer true after this PR is merged.

Also, it'd probably be a good idea to also note that only one of these can be set up, since otherwise a user might think they can schedule as many as they want (I seem to remember at least one topic on DCEmulation where someone was trying to make one of these per controller or something before).

**Review by ljsebald — CHANGES_REQUESTED** [review]

File: kernel/arch/dreamcast/hardware/maple/controller.c
Comment (ljsebald) [inline kernel/arch/dreamcast/hardware/maple/controller.c:147]: Technically, this call here is fine... Since this is in an interrupt handler.

File: kernel/arch/dreamcast/include/dc/maple/controller.h
Comment (QuzarDC) [reply kernel/arch/dreamcast/include/dc/maple/controller.h:214]: Updated the notes on this with both parts. I forgot when I fast-forwarded that Falco had added in the notes about the interrupt context.

**Comment by pcercuei** [comment]
My thoughts on this PR: It would be better to have a *generic* mechanism with an API that allows you to register a handler. The kernel would then create a thread for you, that would call your callback when notified to do so by the main IRQ handler (kind of a mutex+cond, but that would work in IRQ context).

The main IRQ handler would then ACK and mask the IRQ that fired, and unpause the interrupt thread. Only after the registered callback returns should the IRQ be unmasked, and the interrupt thread is back into the "paused" state, waiting for a new interrupt.

Then this PR would just be a matter of registering a button callback with this new threaded interrupt mechanism.

**Comment by pcercuei** [comment]
> We'd still be blocked through by the process of creating non-running threads, which isn't currently supported, so I'll hold this open for feedback from @ljsebald on that for now while working on a draft of a generic version.

Don't we have "genwait" for that? We don't need to create a non-running thread, we can create a running thread that's blocked until the main IRQ handler unblocks it.

**Comment by pcercuei** [comment]
RFC / untested: https://github.com/KallistiOS/KallistiOS/commit/5cdfbf3acf960bf802f08005527c15125dcb6503

**Comment by QuzarDC** [comment]
Just going to close this up for now as it'd have to be rewritten mostly. Will take up the effort in v2.2.0 and see about expanding the functionality as well (allow installing multiple such callbacks. perhaps allow other filtering mechanisms).

**Pull request closed without merging**
